pilot/pkg/model/service.go

```diff
@@ -477,6 +487,23 @@ func ParseServiceKey(s string) (hostname string, ports PortList, labels LabelsCo
 	return
 }
 
+// SubsetKey generates a unique string referencing service instances for a given service name, a subset and a port.
+// The proxy queries Pilot with this key to obtain the list of instances in a subset.
+func BuildSubsetKey(direction TrafficDirection, subsetName, hostname string, port *Port) string {
+	return fmt.Sprintf("%s|%s|%s|%s", direction, subsetName, hostname, port.Name)
+}
+
+// ParseSubsetKey is the inverse of the BuildSubsetKey method
+func ParseSubsetKey(s string) (direction TrafficDirection, subsetName, hostname string, port *Port) {
+	parts := strings.Split(s, "|")
+	// we ignore direction since its typically not used by the consuming functions
```

Comment:
do we want to put in a little safety check on the length of parts?

Reply:
while parsing? or while building? while parsing, it is what it is right? this is coming straight from envoy.